File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>

using namespace std;

int dfs(int index, int sum, const vector<int>& numbers, int target) {
    if (index == numbers.size()) {
        return (sum == target) ? 1 : 0;
    }
    
    return dfs(index + 1, sum + numbers[index], numbers, target) + 
           dfs(index + 1, sum - numbers[index], numbers, target);
}

int solution(vector<int> numbers, int target) {
    return dfs(0, 0, numbers, target);
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    
    int answer = 0;
    
    public int solution(int[] numbers, int target) {
        int sum = 0;
        int current = 0;
        
        dfs(current, sum, numbers, target);
        
        return answer;
    }
    
    void dfs(int current, int sum, int[] numbers, int target) {
        if (current == numbers.length) {
            if (sum == target) {
                answer++;
            }
            return;
        }
        
        dfs(current + 1, sum + numbers[current], numbers, target);
        dfs(current + 1, sum - numbers[current], numbers, target);
    } 
}
